From ebc644a73db37c5f88a6e5cf99357c8ab1ec1bcd Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Fri, 17 Oct 2025 11:24:37 +0800
Subject: [PATCH 225/254] Skip associating AP MLD's link on disallowed channel

The wpa_s->conf->freq_list will be set by ucode, so we can use it to
filter out the BSS(es) that can be connect.

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 wpa_supplicant/sme.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/wpa_supplicant/sme.c b/wpa_supplicant/sme.c
index a366abe41..a12dcee55 100644
--- a/wpa_supplicant/sme.c
+++ b/wpa_supplicant/sme.c
@@ -543,6 +543,14 @@ static void wpas_sme_set_mlo_links(struct wpa_supplicant *wpa_s,
 	for_each_link(bss->valid_links, i) {
 		const u8 *bssid = bss->mld_links[i].bssid;
 
+		if (wpa_s->conf->freq_list &&
+		    !int_array_includes(wpa_s->conf->freq_list,
+					bss->mld_links[i].freq)) {
+			wpa_printf(MSG_DEBUG,
+				   "MLD: Skip the link on disallowed channel");
+			continue;
+		}
+
 		wpa_s->valid_links |= BIT(i);
 		os_memcpy(wpa_s->links[i].bssid, bssid, ETH_ALEN);
 		wpa_s->links[i].freq = bss->mld_links[i].freq;
-- 
2.45.2

